contracts/oracles/TwapOracle.sol

```diff
@@ -196,30 +192,47 @@ contract TwapOracle is OwnableUpgradeable, TwapInterface {
     }
 
     /**
-     * @notice Update new and old observations of lagging window if period elapsed.
+     * @notice Append current Observation and pick equal or just greater than window start timestamp,
+     * The observation which we are using except that we will delete rest of the observationn which are prior to this.
      * @return cumulative price & old observation
      */
-    function pokeWindowValues(TokenConfig memory config) internal returns (uint256, uint256, uint256) {
+    function pokeWindowValues(
+        TokenConfig memory config
+    ) internal returns (uint256, uint256 startCumulativePrice, uint256 startCumulativeTimestamp) {
         uint256 cumulativePrice = currentCumulativePrice(config);
-
-        Observation memory newObservation = newObservations[config.asset];
-
-        // Update new and old observations if elapsed time is greater than or equal to anchor period
-        uint256 timeElapsed = block.timestamp - newObservation.timestamp;
-        if (timeElapsed >= config.anchorPeriod) {
-            oldObservations[config.asset].timestamp = newObservation.timestamp;
-            oldObservations[config.asset].acc = newObservation.acc;
-
-            newObservations[config.asset].timestamp = block.timestamp;
-            newObservations[config.asset].acc = cumulativePrice;
-            emit TwapWindowUpdated(
-                config.asset,
-                newObservation.timestamp,
-                block.timestamp,
-                newObservation.acc,
-                cumulativePrice
-            );
+        uint256 currentTimestamp = block.timestamp;
+        uint256 windowStartTimestamp = currentTimestamp - config.anchorPeriod;
+        Observation[] memory storedObservations = observations[config.asset];
+
+        uint256 storedObservationsLength = storedObservations.length;
+        Observation memory lastObservation = storedObservations[storedObservationsLength - 1];
+
+        //Scenerio when we don't have any observation which falls between (currentTime - anchorPeriod) and currentTime.
+        if (lastObservation.timestamp <= windowStartTimestamp) {
+            startCumulativePrice = lastObservation.acc;
+            startCumulativeTimestamp = lastObservation.timestamp;
+            windowStart[config.asset] = storedObservationsLength - 1;
+        } else {
+            for (uint256 i = windowStart[config.asset]; i < storedObservationsLength; i++) {
```

Comment:
instead of `i++` use `++i`

[Reference](https://yos.io/2021/05/17/gas-efficient-solidity/#tip-12-i-costs-less-gas-compared-to-i-or-i--1)